### ISSUE TITLE
Can't build for ARMV7

### ISSUE THREAD

**Issue by phischmi** [dani-garcia/bitwarden_rs#286]
Hi all,

i'm trying to build the docker image for ARMV7, but whatever i try, the build always fail with the following error:

```
Step 9/33 : RUN npm run sub:init && npm install                                                                                                                                         
 ---> Running in 1fbdbcc0421a                                                                                                                                                           
                                                                                                                                                                                        
> bitwarden-web@2.5.0 sub:init /web-build                                                                                                                                               
> git submodule update --init --recursive                                                                                                                                               
                                                                                                                                                                                        
Submodule 'jslib' (https://github.com/bitwarden/jslib.git) registered for path 'jslib'                                                                                                  
Cloning into '/web-build/jslib'...                                                                                                                                                      
Submodule path 'jslib': checked out '7dcb9b5f8bfd7c65f759b0cc34dfc6abd33fdcd6'                                                                                                          
                                                                                                                                                                                        
> node-sass@4.9.3 install /web-build/node_modules/node-sass                                                                                                                             
> node scripts/install.js                                                                                                                                                               
                                                                                                                                                                                        
Downloading binary from https://github.com/sass/node-sass/releases/download/v4.9.3/linux-arm-57_binding.node                                                                            
Cannot download "https://github.com/sass/node-sass/releases/download/v4.9.3/linux-arm-57_binding.node":                                                                                 
                                                                                                                                                                                        
HTTP error 404 Not Found                                                                                                                                                                
                                                                                                                                                                                        
Hint: If github.com is not accessible in your location                                                                                                                                  
      try setting a proxy via HTTP_PROXY, e.g.                                                                                                                                          
                                                                                                                                                                                        
      export HTTP_PROXY=http://example.com:1234                                                                                                                                         
                                                                                                                                                                                        
or configure npm proxy via                                                                                                                                                              
                                                                                                                                                                                        
      npm config set proxy http://example.com:8080                                                                                                                                      
                                                                                                                                                                                        
> ngx-infinite-scroll@6.0.1 postinstall /web-build/node_modules/ngx-infinite-scroll                                                                                                     
> opencollective postinstall                                                                                                                                                            
                                                                                                                                                                                        
                                                                                                                                                                                        
     *** Thank you for using ngx-infinite-scroll! ***                                                                                                                                   
                                                                                                                                                                                        
Please consider donating to our open collective                                                                                                                                         
     to help us maintain this package.                                                                                                                                                  
                                                                                                                                                                                        
  https://opencollective.com/ngx-infinite-scroll/donate                                                                                                                                 
                                                                                                                                                                                        
                    ***                                                                                                                                                                 
                                                                                                                                                                                        
                                                                                                                                                                                        
> node-sass@4.9.3 postinstall /web-build/node_modules/node-sass                                                                                                                         
> node scripts/build.js                                                                                                                                                                 
                                                                                                                                                                                        
Building: /usr/local/bin/node /web-build/node_modules/node-gyp/bin/node-gyp.js rebuild --verbose --libsass_ext= --libsass_cflags= --libsass_ldflags= --libsass_library=                 
gyp info it worked if it ends with ok                                                                                                                                                   
gyp verb cli [ '/usr/local/bin/node',                                                                                                                                                   
gyp verb cli   '/web-build/node_modules/node-gyp/bin/node-gyp.js',                                                                                                                      
gyp verb cli   'rebuild',                                                                                                                                                               
gyp verb cli   '--verbose',                                                                                                                                                             
gyp verb cli   '--libsass_ext=',                                                                                                                                                        
gyp verb cli   '--libsass_cflags=',                                                                                                                                                     
gyp verb cli   '--libsass_ldflags=',                                                                                                                                                    
gyp verb cli   '--libsass_library=' ]                                                                                                                                                   
gyp info using node-gyp@3.8.0                                                                                                                                                           
gyp info using node@8.14.0 | linux | arm                                                                                                                                                
gyp verb command rebuild []                                                                                                                                                             
gyp verb command clean []                                                                                                                                                               
gyp verb clean removing "build" directory                                                                                                                                               
gyp verb command configure []                                                                                                                                                           
gyp verb check python checking for Python executable "python2" in the PATH                                                                                                              
gyp verb `which` failed Error: not found: python2                                                                                                                                       
gyp verb `which` failed     at getNotFoundError (/web-build/node_modules/which/which.js:13:12)                                                                                          
gyp verb `which` failed     at F (/web-build/node_modules/which/which.js:68:19)                                                                                                         
gyp verb `which` failed     at E (/web-build/node_modules/which/which.js:80:29)                                                                                                         
gyp verb `which` failed     at /web-build/node_modules/which/which.js:89:16                                                                                                             
gyp verb `which` failed     at /web-build/node_modules/isexe/index.js:42:5                                                                                                              
gyp verb `which` failed     at /web-build/node_modules/isexe/mode.js:8:5                                                                                                                
gyp verb `which` failed     at FSReqWrap.oncomplete (fs.js:152:21)                                                                                                                      
gyp verb `which` failed  python2 { Error: not found: python2                                                                                                                            
gyp verb `which` failed     at getNotFoundError (/web-build/node_modules/which/which.js:13:12)                                                                                          
gyp verb `which` failed     at F (/web-build/node_modules/which/which.js:68:19)                                                                                                         
gyp verb `which` failed     at E (/web-build/node_modules/which/which.js:80:29)                                                                                                         
gyp verb `which` failed     at /web-build/node_modules/which/which.js:89:16                                                                                                             
gyp verb `which` failed     at /web-build/node_modules/isexe/index.js:42:5                                                                                                              
gyp verb `which` failed     at /web-build/node_modules/isexe/mode.js:8:5                                                                                                                
gyp verb `which` failed     at FSReqWrap.oncomplete (fs.js:152:21)                                                                                                                      
gyp verb `which` failed   stack: 'Error: not found: python2\n    at getNotFoundError (/web-build/node_modules/which/which.js:13:12)\n    at F (/web-build/node_modules/which/which.js:68
:19)\n    at E (/web-build/node_modules/which/which.js:80:29)\n    at /web-build/node_modules/which/which.js:89:16\n    at /web-build/node_modules/isexe/index.js:42:5\n    at /web-buil
d/node_modules/isexe/mode.js:8:5\n    at FSReqWrap.oncomplete (fs.js:152:21)',                                                                                                          
gyp verb `which` failed   code: 'ENOENT' }                                                                                                                                              
gyp verb check python checking for Python executable "python" in the PATH                                                                                                               
gyp verb `which` failed Error: not found: python                                                                                                                                        
gyp verb `which` failed     at getNotFoundError (/web-build/node_modules/which/which.js:13:12)                                                                                          
gyp verb `which` failed     at F (/web-build/node_modules/which/which.js:68:19)                                                                                                         
gyp verb `which` failed     at E (/web-build/node_modules/which/which.js:80:29)                                                                                                         
gyp verb `which` failed     at /web-build/node_modules/which/which.js:89:16                                                                                                             
gyp verb `which` failed     at /web-build/node_modules/isexe/index.js:42:5                                                                                                              
gyp verb `which` failed     at /web-build/node_modules/isexe/mode.js:8:5                                                                                                                
gyp verb `which` failed     at FSReqWrap.oncomplete (fs.js:152:21)                                                                                                                      
gyp verb `which` failed  python { Error: not found: python                                                                                                                              
gyp verb `which` failed     at getNotFoundError (/web-build/node_modules/which/which.js:13:12)                                                                                          
gyp verb `which` failed     at F (/web-build/node_modules/which/which.js:68:19)                                                                                                         
gyp verb `which` failed     at E (/web-build/node_modules/which/which.js:80:29)                                                                                                         
gyp verb `which` failed     at /web-build/node_modules/which/which.js:89:16                                                                                                             
gyp verb `which` failed     at /web-build/node_modules/isexe/index.js:42:5                                                                                                              
gyp verb `which` failed     at /web-build/node_modules/isexe/mode.js:8:5                                                                                                                
gyp verb `which` failed     at FSReqWrap.oncomplete (fs.js:152:21)                                                                                                                      
gyp verb `which` failed   stack: 'Error: not found: python\n    at getNotFoundError (/web-build/node_modules/which/which.js:13:12)\n    at F (/web-build/node_modules/which/which.js:68:
19)\n    at E (/web-build/node_modules/which/which.js:80:29)\n    at /web-build/node_modules/which/which.js:89:16\n    at /web-build/node_modules/isexe/index.js:42:5\n    at /web-build
/node_modules/isexe/mode.js:8:5\n    at FSReqWrap.oncomplete (fs.js:152:21)',                                                                                                           
gyp verb `which` failed   code: 'ENOENT' }                                                                                                                                              
gyp ERR! configure error                                                                                                                                                                
gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.                                                                                       
gyp ERR! stack     at PythonFinder.failNoPython (/web-build/node_modules/node-gyp/lib/configure.js:484:19)                                                                              
gyp ERR! stack     at PythonFinder.<anonymous> (/web-build/node_modules/node-gyp/lib/configure.js:406:16)                                                                               
gyp ERR! stack     at F (/web-build/node_modules/which/which.js:68:16)                                                                                                                  
gyp ERR! stack     at E (/web-build/node_modules/which/which.js:80:29)                                                                                                                  
gyp ERR! stack     at /web-build/node_modules/which/which.js:89:16                                                                                                                      
gyp ERR! stack     at /web-build/node_modules/isexe/index.js:42:5                                                                                                                       
gyp ERR! stack     at /web-build/node_modules/isexe/mode.js:8:5                                                                                                                         
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:152:21)                                                                                                                               
gyp ERR! System Linux 4.14.69-odroidxu4                                                                                                                                                 
gyp ERR! command "/usr/local/bin/node" "/web-build/node_modules/node-gyp/bin/node-gyp.js" "rebuild" "--verbose" "--libsass_ext=" "--libsass_cflags=" "--libsass_ldflags=" "--libsass_lib
rary="                                                                                                                                                                                  
gyp ERR! cwd /web-build/node_modules/node-sass                                                                                                                                          
gyp ERR! node -v v8.14.0                                                                                                                                                                
gyp ERR! node-gyp -v v3.8.0                                                                                                                                                             
gyp ERR! not ok                                                                                                                                                                         
Build failed with error code: 1                                                                                                                                                         
npm WARN bitwarden-web@2.5.0 No repository field.                                                                                                                                       
npm WARN bitwarden-web@2.5.0 No license field.                                                                                                                                          
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):                                                                                                 
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"arm"})                       
                                                                                                                                                                                        
npm ERR! code ELIFECYCLE                                                                                                                                                                
npm ERR! errno 1                                                                                                                                                                        
npm ERR! node-sass@4.9.3 postinstall: `node scripts/build.js`                                                                                                                           
npm ERR! Exit status 1                                                                                                                                                                  
npm ERR!                                                                                                                                                                                
npm ERR! Failed at the node-sass@4.9.3 postinstall script.                                                                                                                              
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.                                                                                      
                                                                                                                                                                                        
npm ERR! A complete log of this run can be found in:                                                                                                                                    
npm ERR!     /root/.npm/_logs/2018-12-13T09_32_14_077Z-debug.log                                                                                                                        
The command '/bin/sh -c npm run sub:init && npm install' returned a non-zero code: 1
```

Is there anything i can do to make this work?

**Comment by mprasil** [comment]
Are you actually building on armv7 device? There's `Dockerfile.armv7` that's meant to be built on x86 device, but it cross-compiles to armv7.

**Comment by phischmi** [comment]
I'm building on an Odroid HC2 which is armv7

**Comment by mprasil** [comment]
It seems like this might be an upstream issue, I'm not even sure if Vault actually can be built on armv7. Perhaps you can overcome this by using our image for the vault data? Just replace the Vault build stage with:

```
FROM mprasil/bitwarden as vault
```
Then delete the rest up to the binary build image stage. (that starts with `FROM rust as build`)

**Comment by phischmi** [comment]
> It seems like this might be an upstream issue, I'm not even sure if Vault actually can be built on armv7. Perhaps you can overcome this by using our image for the vault data? Just replace the Vault build stage with:
> 
> ```
> FROM mprasil/bitwarden as vault
> ```
> Then delete the rest up to the binary build image stage. (that starts with `FROM rust as build`)

That seems to work :) Thanks!

**Comment by mprasil** [comment]
It's a bit of a workaround, but I guess there isn't much we can do here as the code is all upstream. I'm going to close this now, glad that it worked for you.

**Comment by thiscantbeserious** [comment]
The issue above seems like one we've faced as well, a quiet ugly one at it.

node-sass is basically just a binding to LibSass - a native c++ sass compiler, node-sass (seriously) tries to look for Python and C++ binaries and/or tries to pull them to compile the underlying LibSass for your local environment to be able to comple Sass to CSS via the binding node-sass provides - resulting in one heck of a dependency mess that's hard to manage.

I'll create an issue in the official bitwarden repo for the webvault suggesting to switch to https://www.npmjs.com/package/sass which is a native JS SASS compiler that doesn't require any hacky dependencies out of its domain (JS).

The performance gains from node-sass are little compared to its actualy usecase from my experience (compiling about a hundred sass files for a frontend-framework that we've created).
